cyberark/secrets-provider-for-k8s
# Add Annotation parsing and Config from Annotations functions

### What does this PR do?

This PR allows for configuring Secrets Provider with Pod annotations. Annotation settings are prioritized over environment variable settings, but envvars can still be used to configure SP to maintain backwards compatibility.

#### Adds `annotations_parser.go` and `annotation_parser_test.go` to package `secrets/annotations`, including:
- `AnnotationsFromFile(path string) (map[string]string, error)`
- `ParseAnnotations(annotationsFile io.Reader) (map[string]string, error)`
- **Test Cases**
  - A valid annotations file is parsed successfully
  - Malformed annotations file lines result in an error indicating the line number and desired format

#### Adds functions to `secrets/config` package:
- `ValidateAnnotations(annotations map[string]string) ([]error, []error)`
  - Confirms annotation keys & value types, and confirms value for those annotations with enumerated value options
  - Returns lists of Error-level and Info-level logs
  - **Test Cases**
    - Properly formatted annotations return empty arrays
    - Confirm Info log if for unrecognized annotations with and without `conjur.org/` prefix
    - Confirm Error log for annotations given improper value type
    - Confirm Error log for annotations with explicit input ranges given invalid values
- `GatherSecretsProviderSettings(annotations map[string]string) map[string]string`
  - Returns an map containing annotations and envvars required from Secrets Provider config
  - **Test Cases**
    - If the annotations map is empty, the resulting map contains envvars used in SP config
    - If envvars are not set, the resulting map contains contains the annotations used in SP config
    - If both are present, the resulting map is all envvars and annotations used in SP config
    - Annotations and envvars not used in SP config are excluded
- `ValidateSecretsProviderSettings(envAndAnnots map[string]string) ([]error, []error)`
  - Confirm that the provided annotations and envvars lend to a valid SP config
  - Input validation on annotation keys and values
  - Returns Error and Info level log messages. An empty Error array means validation has succeeded.
  - **Test Cases**
    - Given a valid configuration of annotations, no errors are returned
    - Given a valid configuration of envvars, no errors are returned
    - Confirm that `MY_POD_NAMESPACE` is required in either case
    - Confirm Error log if StoreType is not set
    - Confirm Error log is Push-to-File is being configured with envvars
    - Confirm Error log if K8s Secrets mode is missing required K8s secrets
    - Confirm Info log if values are provided to equivalent annotation and envvar
- `NewConfig(settings map[string]string) *Config`
  - Create a new SP config from a validated map of annotations and SP Config envvars
  - **Test Cases**
    - A valid map of annotation-based settings results in a valid Config
    - A valid map of envvar-based settings results in a valid Config
    - A valid map of both annotation and envvar settings defers to annotation values

#### New functions are used in SP main function to create a new SP Config

#### Acceptance Criteria Review
- [x] An annotation parser exists for extracting arbitrary annotations from a downward API annotation file. This should include a test case to validate this behavior.
- [x] Annotations can be used instead of environment variables. This should include a test case to validate this behavior.
- [x] Annotations take precedence over environment variables. This should include a test case to validate this behavior.
- [x] Unknown Conjur annotation causes INFO level log
- [x] Multiline annotation values are parsed correctly


### What ticket does this PR close?
ONYX-11827
Closes #330 

### Checklists

#### Change log

#### Test coverage

#### Documentation
